Support range based header matching for request routing.

kavyako opened the pull request:
v2 api changes: Use oneof to specify header match options, based on value, regex or range (specific to int64 values)
The existing value and regex fields will be deprecated. USe the header_match_specfier oneof instead.
The range_match option has a low_key and high_key. Route match succeeds if the header value lies within this range (inclusive).
v1 api: Add a range_match object to the route headers json. Presence of this object indicates range based route match.

https://github.com/envoyproxy/envoy/issues/2418

kavyako commented:
@mattklein123 Here is the PR for envoyproxy/envoy#2418

mattklein123 commented:
Also FYI you will need to fix DCO. See https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#dco-sign-your-work.

htuch commented on envoy/api/v2/route/route.proto:
```diff
@@ -768,6 +768,17 @@ message RateLimit {
+// Header route matching will be performed if the header's value lies within this range (inclusive).
+message HeaderValueRange {
+  // Low key, inclusive
+  google.protobuf.Int64Value low_key = 1;
```
Do we need a wrapped type here? Is the idea that if it's not specified, this is an open ended range (lower is min int64 if missing, upper is max int64)?

kavyako replied:
IMO the low and high could be "required" fields, no default values will be assumed. Being more explicit will avoid confusion. What do you say? 
Regarding using wrapped types, are there any disadvantages? New to protos here so I can change this to int64 if that is the recommended way.

wora replied:
HeaderValueRange is already optional. The low and high can be required. By the way, what "key" means here.

If the comment says, lower bound and upper bound, the field name should read the same. Please avoid asking readers to correlate "lower bound" and "low key".

mattklein123 replied:
+1, I would also use primitive types here. IMO default of 0 and 0 is fine.

wora replied:
Please see my other comment about defining a more generic type. This type is so commonly used, we should promote it to the top level.

```
message Range {
  double start;
  double end;
}
```

kavyako replied:
@wora top level => some common proto file? I see core/base.proto. Is that the place for such generic types?
Based on Matt's other comment, to avoid double conversions, shall we stick to int64 then?

```
message Int64Range {
  required int64 start;
  required int64 end;
}
```

wora replied:
For types that are never going to change, I suggest we put them at the very top level, e.g. /envoy/type/range.proto. We can put Int64Range there first, add more if needed.

htuch commented on envoy/api/v2/route/route.proto:
```diff
@@ -768,6 +768,17 @@ message RateLimit {
+  google.protobuf.Int64Value low_key = 1;
+
+  // High key, inclusive
+  google.protobuf.Int64Value high_key = 2;
```
Nit: prefer just making these `high/low` or `high_value/low_value`. Key here is confusing.

kavyako replied:
yes, I ll change this to low/high.

htuch commented on docs/root/api-v1/route_config/route.rst:
```diff
@@ -338,7 +338,8 @@ Headers
     "name": "...",
     "value": "...",
-    "regex": "..."
+    "regex": "...",
```
Do we need this in the v1 API? If you don't have a specific need for this, we prefer to feature freeze v1 and not add new things.

kavyako replied:
We are using the v1 APIs internally and it will be some time before we switch to v2.

htuch replied:
OK, that's fine for now. We are likely to start pushing harder on this over time so we can turn down v1 at some point, but we haven't announced anything yet and there will be at least a quarter or two notice.

htuch commented on envoy/api/v2/route/route.proto:
```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
   // * The regex *\d{3}* does not match the value *1234*
   // * The regex *\d{3}* does not match the value *123.456*
-  google.protobuf.BoolValue regex = 3;
+  google.protobuf.BoolValue regex = 3 [deprecated = true];
```
We shouldn't deprecate until the new one is ready yet :)

kavyako replied:
will remove the tag.

htuch commented on envoy/api/v2/route/route.proto:
```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
+
+  // [#not-implemented-hide:]
+  // Specifies how the header match will be performed to route the request.
+  // If header_match_specifier is absent, a request that has the name header will match, regardless
```
"a request that has :ref:`header <internal link back to header field>` will match". I know this problem was there before, this is a good place to cleanup.

htuch commented on envoy/api/v2/route/route.proto:
```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
+    // If specified, header match will be performed based on the value of the header.
+    string value_match = 4;
+
+    // [#not-implemented-hide:]
```
If you have an outer scope not-implemented-hide, it applies to the entire subtree in AST.

kavyako replied:
ok, will remove these tags.

kavyako replied:
@htuch  looks like that doesn't work for oneofs. I got a warning/error while building documentation:
 envoy_api_msg_route.int64range (if the link has no caption the label must precede a section header)

wora commented on envoy/api/v2/route/route.proto:
```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
+    // If specified, header match will be performed based on range.
+    // The rule will match if the request header value is within this range.
+    // This only applies to integer header values.
+    HeaderValueRange range_match = 6;
```
The type name can be RangeMatch. The match can be applied to any value. It doesn't need to be from header.

htuch replied:
`Interval` might be another name. It doesn't need to be used in a match capacity..

wora replied:
Interval often misleads to time. Range is more accurate in this context.

htuch replied:
OK, then simply `Range`?

wora replied:
Google API Design Guide has a specific section on the range representation. We should always use [start, end) to represent a range. The reason is to avoid English debate like, [lower, upper], [lower, high], [first, last]. It is better to have a single universal representation instead of many locally optimized one.

Do we really care about int64 vs double? Once someone starts to deal with numbers bigger than a billion, it is better to use double, 1000000000000 is much worse to write than 1e12. I don't think there is any business reason to limit the header value to integers.

mattklein123 replied:
I would strongly recommend defining an `IntegerRange` or `Int64Range` and using it for this feature. In the data path there is no reason to be performing double conversions in this case.

rshriram replied:
I would suggest naming the field as "integer_range" instead of just range. I would also like to see the use case as @htuch said..

kavyako replied:
@rshriram some explanation here https://github.com/envoyproxy/envoy/issues/2418
A very specific use case: Service Fabric supports a concept of [ranged partitioning scheme ](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-concepts-partitioning#ranged-partitioning-scheme) for stateful services. 
An incoming request will be routed to a given partition of the stateful service if the partition key is in the specified range for that partition.

rshriram replied:
This is cool! Could you just stick this in the comments for this field? It will greatly help in the documentation when people look at this option, and get a sense for what the real world usage is.

rshriram commented on envoy/api/v2/route/route.proto:
```diff
@@ -803,7 +814,36 @@ message HeaderMatcher {
+  oneof header_match_specifier {
+    // [#not-implemented-hide:]
+    // If specified, header match will be performed based on the value of the header.
+    string value_match = 4;
```
nit: exact_match, prefix_match, regex_match (or get rid of _match). Prefix match is pretty helpful as well

kavyako replied:
@rshriram  I went the _match suffix to avoid clashes with the existing value, regex fields in this message, which will be deprecated once people switch to this oneof match specifier. and named this range_match for consistency. 
Can change the value_match to exact_match so that we have scope to implement prefix_match later (I am not planning to add that as part of this PR).

kavyako commented:
@mattklein123 : Added DCO and fixed the file mode changes.
@wora : I have added a new type range.proto under envoy/type/v2 (all the proto s seem to be under a v2 sub folder).
Went with sint64 for the range start and end values. Proto3 document recommends using sint as it more efficiently encodes negative numbers.

yet to consume/verify these changes in the envoy repo.

kavyako commented on envoy/type/v2/BUILD:
```diff
@@ -0,0 +1,14 @@
+api_proto_library(
+    name = "range",
+    srcs = ["range.proto"],
+    visibility = ["//visibility:public"],
```
What should the visibility be set to? 

//envoy/api/v2:friends gives the error:
ERROR: /source/envoy/api/v2/route/BUILD:5:1: Target '//envoy/type/v2:range' is not visible from target '//envoy/api/v2/route:route'. Check the visibility declaration of the former target if you think the dependency is legitimate

htuch replied:
Just leave it empty, the default in https://github.com/envoyproxy/data-plane-api/blob/master/bazel/api_build_system.bzl#L81 should apply unless you have specific need to restrict visibility as done elsewhere.

kavyako replied:
Empty does not work,  default => private, I get the same error as mentioned above. This type is referenced in envoy/api/v2/route

htuch commented on envoy/type/v2/range.proto:
```diff
@@ -0,0 +1,20 @@
+package envoy.type.v2;
+option go_package = "v2";
+
+import "validate/validate.proto";
```
Not needed.

wora commented on envoy/type/v2/range.proto:
```diff
@@ -0,0 +1,20 @@
+
+// Specifies the signed int64 start and end of the range using half-open interval semantics [start,
+// end).
+message SInt64Range {
```
You don't nee S here. It doesn't mean anything different from Int64.

wora commented on envoy/type/v2/range.proto:
```diff
@@ -0,0 +1,20 @@
+// end).
+message SInt64Range {
+  // start of the range (inclusive)
+  sint64 start = 1;
```
Google APIs never use sint64, since negative values are rarely used in practice. You are paying more encoding/decoding cost to save wire format for negative value. At Google, even we process billions of requests per second, we sill favor simplicity than wire cost.

wora commented on envoy/type/v2/range.proto:
```diff
@@ -0,0 +1,20 @@
+
+// [#proto-status: draft]
+
+package envoy.type.v2;
```
Please just use envoy.type. Any package that ends with `.type` promises never have any breaking change, like `google.type`. In some cases, we should bite the bullet to make hard promise.

This is one reason I encourage Envoy and Istio to take dependency on google.type and google.rpc. Google APIs promise not to make any breaking change to them.

htuch commented on envoy/type/BUILD:
```diff
@@ -0,0 +1,14 @@
+api_proto_library(
+    name = "range",
+    srcs = ["range.proto"],
+    visibility = ["//visibility:public"],
```
I think this can just be elided and it will work by magic with the default.

kavyako replied:
@htuch Removing the visibility line or passing empty does not work.

The default in https://github.com/envoyproxy/data-plane-api/blob/master/bazel/api_build_system.bzl#L81 
is private.

ERROR: /source/envoy/api/v2/route/BUILD:5:1: Target '//envoy/type/v2:range' is not visible from target '//envoy/api/v2/route:route'. Check the visibility declaration of the former target if you think the dependency is legitimate

htuch commented on envoy/type/range.proto:
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
```
I think what you have here is good enough for final status.

zuercher commented on envoy/api/v2/route/route.proto:
```diff
@@ -815,6 +816,38 @@ message HeaderMatcher {
+    // [#not-implemented-hide:]
+    // If specified, header match will be performed based on range.
+    // The rule will match if the request header value is within this range.
+    // This only applies to integer header values.
```
Can you be explicitly about what happens when the header's value cannot be converted to an integer? (e.g. Is it treated as zero or does it never match? Are floating point numbers truncated to integers or treated as non-integer strings?)

wora commented:
The range.proto looks really nice. It is simple and has well defined
semantics. It settles the direction for similar data types for Envoy in the
long term.

FWIW, we sometimes have to spend time to design basic data types that are
committed not to have breaking change. The long term stability will save
data conversion cost at runtime.

/lgtm
